### PR TITLE
delete resources in workspace when workspace is deleted

### PR DESCRIPTION
**Pull request by qiujian16**
Signed-off-by: Jian Qiu <jqiu@redhat.com>

## Summary
basically reuse the namespace deletion code to delete all resources in a workspace when workspace is deleting. Still need to add tests

## Related issue(s)

Fixes https://github.com/kcp-dev/kcp/issues/527

### CODE REVIEW

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+
+	c.deleter = deletion.NewWorkspacedResourcesDeleter(kcpClient, metadataClient, discoverResourcesFn, "ws-finalize")
+	c.workspaceSynced = workspaceInformer.Informer().HasSynced
+
+	return c
+}
+
+type GCController struct {
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go:64]: this is not really GC, but just namespace and object creation, right? Would just call it `clusterworkspacedeletion` as a package. 

Also, don't mix it with the existing clusterworkspace controller. Deletion is separate IMO.

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+				hasFinalizer = true
+				break
+			}
+		}
+		if !hasFinalizer {
+			workspace.Finalizers = append(workspace.Finalizers, deletion.WorkspaceFinalizer)
+			_, err := c.kcpClient.Cluster(logicalcluster.From(workspace)).TenancyV1alpha1().ClusterWorkspaces().Update(ctx, workspace, metav1.UpdateOptions{})
+			return err
+		}
+	}
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go:172]: return? You don't want to run into the deletion below, do you?

File: pkg/reconciler/tenancy/clusterworkspace/deletion/status_condition_util.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+)
+
+// NamespaceConditionUpdater interface that translates namespace deleter errors
+// into namespace status conditions.
+type WorkspaceConditionUpdater interface {
+	ProcessDiscoverResourcesErr(e error)
+	ProcessGroupVersionErr(e error)
+	ProcessDeleteContentErr(e error)
+	Update(*tenancyv1alpha1.ClusterWorkspace) bool
+}
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/deletion/status_condition_util.go:38]: can you follow the patterns for updating conditions in pkg/reconciler/pis/apibindings? This code here feels different in style.

File: pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go [hunk 4]
```diff
@@ -0,0 +1,536 @@
+/*
+Copyright 2021 The KCP Authors.
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go:2]: 2022

File: pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go [hunk 5]
```diff
@@ -0,0 +1,536 @@
+/*
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go:1]: how much of this file is from namespace deletion?

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2021 The KCP Authors.
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go:2]: 2022

File: pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/metadata"
+	"k8s.io/client-go/util/retry"
+)
+
+const WorkspaceFinalizer = "workspace-finalizer"
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go:41]: Would rather use something with `tenancy.kcp.dev`.

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+		klog.Infof("Workspace has been deleted %v", key)
+		return nil
+	}
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("unable to retrieve workspace %v from store: %v", key, err))
+		return err
+	}
+
+	workspace = workspace.DeepCopy()
+	hasFinalizer := false
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go:159]: move into the if clause

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go (hunk 1 above)
Comment (qiujian16) [reply pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_gc_controller.go]: renamed

File: pkg/reconciler/tenancy/clusterworkspace/deletion/status_condition_util.go (hunk 3 above)
Comment (qiujian16) [reply pkg/reconciler/tenancy/clusterworkspace/deletion/status_condition_util.go]: updated

File: pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go (hunk 5 above)
Comment (qiujian16) [reply pkg/reconciler/tenancy/clusterworkspace/deletion/workspace_resource_deletor.go]: I think 70%. I remove the opCache part, and change how conditions are updated. Add a comment, and also add kubernetes copyright in this file.

**Comment by qiujian16** [comment]
@sttts PTAL on this PR again. Thanks!

File: pkg/admission/namespacelifecycle/admission.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	org, hasParent := clusterName.Parent()
+	if !hasParent {
+		return l.legacyNamespaceLifecycle.Admit(ctx, a, o)
+	}
+
+	workspaceKey := clusters.ToClusterAwareKey(org, clusterName.Base())
+
+	workspace, err := l.workspaceLister.Get(workspaceKey)
```
Comment (sttts) [inline pkg/admission/namespacelifecycle/admission.go:108]: this won't scale. I would rather prefer some way to decide locally, e.g. by looking at the request user, to decide whether we want to allow deletion.

Background: as soon as we get sharding you won't have a `workspaceLister` anymore that holds all workspaces, only those on the same shard.

File: pkg/admission/namespacelifecycle/admission.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+// ValidateInitialization implements the InitializationValidator interface.
+func (l *workspaceNamespaceLifecycle) ValidateInitialization() error {
+	if err := l.legacyNamespaceLifecycle.ValidateInitialization(); err != nil {
+		return err
+	}
+
+	if err := l.namespaceLifecycle.ValidateInitialization(); err != nil {
+		return err
+	}
+	return nil
```
Comment (sttts) [inline pkg/admission/namespacelifecycle/admission.go:147]: no check for the kcp informer?

File: pkg/admission/plugins.go [hunk 11]
```diff
@@ -45,10 +45,12 @@ import (
 	"github.com/kcp-dev/kcp/pkg/admission/clusterworkspaceshard"
 	"github.com/kcp-dev/kcp/pkg/admission/clusterworkspacetype"
 	"github.com/kcp-dev/kcp/pkg/admission/clusterworkspacetypeexists"
+	"github.com/kcp-dev/kcp/pkg/admission/namespacelifecycle"
```
Comment (sttts) [inline pkg/admission/plugins.go:48]: maybe alias as `kcpnamespacelifecycle`, to make clear it is our version

File: pkg/admission/namespacelifecycle/admission.go (hunk 9 above)
Comment (sttts) [reply pkg/admission/namespacelifecycle/admission.go:108]: Or let's say this admission plugin must not fail normal namespace deletion when the shard with the workspace object is down. During workspace deletion though, failing in that case is ok.

File: pkg/apis/tenancy/v1alpha1/types.go [hunk 12]
```diff
@@ -190,6 +190,23 @@ const (
...
+	WorkspaceDeletionGVParsingSuccess conditionsv1alpha1.ConditionType = "WorkspaceDeletionGVParsingSuccess"
+
+	// WorkspaceDeletionContentSuccess represents the status that all resources in the workspace is deleting
+	WorkspaceDeletionContentSuccess conditionsv1alpha1.ConditionType = "WorkspaceDeletionContentSuccess"
+
+	// WorkspaceContentClean represents the status that all resources in the workspace is cleaned.
+	WorkspaceContentClean conditionsv1alpha1.ConditionType = "WorkspaceContentClean"
+
+	// WorkspaceFinalizersClean represents the status that finalizers of all resources in the workspace is cleaned.
+	WorkspaceFinalizersClean conditionsv1alpha1.ConditionType = "WorkspaceFinalizersClean"
```
Comment (sttts) [inline pkg/apis/tenancy/v1alpha1/types.go:209]: these are many conditions. Can we combine them into one, but with many reasons?

File: pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+	klog.Infof("Queueing workspace %q", key)
+	c.queue.Add(key)
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting ClusterWorkspace GC controller")
+	defer klog.Info("Shutting down ClusterWorkspace GC controller")
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go:93]: update controller name

File: pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+	copiedFinalizers := []string{}
+	for i := range workspace.Finalizers {
+		if workspace.Finalizers[i] == deletion.WorkspaceFinalizer {
+			continue
+		}
+		copiedFinalizers = append(copiedFinalizers, workspace.Finalizers[i])
+	}
+	if len(workspace.Finalizers) != len(copiedFinalizers) {
+		workspace.Finalizers = copiedFinalizers
+		_, err := c.kcpClient.Cluster(logicalcluster.From(workspace)).TenancyV1alpha1().ClusterWorkspaces().Update(ctx, workspace, metav1.UpdateOptions{})
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go:241]: patch?

File: pkg/admission/namespacelifecycle/admission.go (hunk 9 above)
Comment (qiujian16) [reply pkg/admission/namespacelifecycle/admission.go:116]: good point! updated.

File: pkg/apis/tenancy/v1alpha1/types.go (hunk 12 above)
Comment (qiujian16) [reply pkg/apis/tenancy/v1alpha1/types.go]: reduced to two conditions WorkspaceContentClean and WorkspaceContentDeleteSuccess

File: test/e2e/framework/expect.go [hunk 15]
```diff
@@ -51,7 +51,7 @@ type Expecter interface {
 
 // NewExpecter creates a informer-driven registry of expectations, which will
 // be triggered on every event that the informer ingests.
-func NewExpecter(informer cache.SharedIndexInformer) *expectationController {
+func NewExpecter(informer cache.SharedIndexInformer) Expecter {
```
Comment (ncdc) [inline test/e2e/framework/expect.go:54]: I'm curious why you made this change?

Comment (qiujian16) [reply test/e2e/framework/expect.go:54]: I think this should return an interface but not a private class? I need to call ExpectBefore method of this interface https://github.com/kcp-dev/kcp/pull/913/files#diff-d37089ee294b6de484fc1d0d69e3ce8357b20367d45f1336df40e23a2461da7bR67

Comment (ncdc) [reply test/e2e/framework/expect.go:54]: See https://github.com/golang/go/wiki/CodeReviewComments#interfaces - it's better to return the concrete type

Comment (qiujian16) [reply test/e2e/framework/expect.go]: gotcha, updated :)

File: pkg/admission/namespacelifecycle/admission.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+func (l *workspaceNamespaceLifecycle) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+
+	// call legacy namespace lifecycle at first
+	admissionErr := l.legacyNamespaceLifecycle.Admit(ctx, a, o)
+
+	// Only handle the case when namespace deletion is forbidden
+	if !apierrors.IsForbidden(admissionErr) {
+		return admissionErr
+	}
+
```
Comment (sttts) [inline pkg/admission/namespacelifecycle/admission.go:99]:
```suggestion
// when the workspace is being deleted, we want to allow deletion of all workspaces. Hence, after 
// legacyNamespaceLifecycle has forbidden the deletion, we give it a second chance be delegating to
// namespaceLifecycle which does not protect any namespace beyond normal life-cycle.
```

File: pkg/admission/namespacelifecycle/admission.go [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+	if !hasParent {
+		return admissionErr
+	}
+
+	workspaceKey := clusters.ToClusterAwareKey(org, clusterName.Base())
+
+	workspace, err := l.workspaceLister.Get(workspaceKey)
+	// The shard hosting the workspace could done,
+	// just return error from legacy namespace lifecycle addmission in this case
+	if err != nil {
```
Comment (sttts) [inline pkg/admission/namespacelifecycle/admission.go:119]: do we want to handle 404 differently? A 404 is a definitive response, i.e. the workspace is really gone. Maybe we want to keep deleting the namespace then?

File: pkg/admission/plugins.go [hunk 18]
```diff
@@ -78,10 +80,11 @@ func RegisterAllKcpAdmissionPlugins(plugins *admission.Plugins) {
 	clusterworkspacetypeexists.Register(plugins)
 	apiresourceschema.Register(plugins)
 	apibinding.Register(plugins)
+	kcpnamespacelifecycle.Register(plugins)
 }
 
 var defaultOnPluginsInKcp = sets.NewString(
-	lifecycle.PluginName,              // NamespaceLifecycle
+	kcpnamespacelifecycle.PluginName,  // NamespaceLifecycle
```
Comment (sttts) [inline pkg/admission/plugins.go:87]: WorkspaceNamespaceLifecycle

File: pkg/apis/tenancy/v1alpha1/types.go [hunk 19]
```diff
@@ -190,6 +190,12 @@ const (
 	// WorkspaceShardValidReasonShardNotFound reason in WorkspaceShardValid condition means that the
 	// referenced ClusterWorkspaceShard object got deleted.
 	WorkspaceShardValidReasonShardNotFound = "ShardNotFound"
+
+	// WorkspaceDeletionContentSuccess represents the status that all resources in the workspace is deleting
+	WorkspaceDeletionContentSuccess conditionsv1alpha1.ConditionType = "WorkspaceDeletionContentSuccess"
+
+	// WorkspaceContentClean represents the status that all resources in the workspace is cleaned.
+	WorkspaceContentClean conditionsv1alpha1.ConditionType = "WorkspaceContentClean"
```
Comment (sttts) [inline pkg/apis/tenancy/v1alpha1/types.go:198]: Clean = Deleted?

File: test/e2e/reconciler/clusterworkspacedeletion/controller_test.go [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+					if !conditions.IsTrue(workspace, tenancyv1alpha1.WorkspaceDeletionContentSuccess) {
+						return false, fmt.Errorf("expect true status for condition %s", tenancyv1alpha1.WorkspaceDeletionContentSuccess)
+					}
+
+					if !conditions.IsFalse(workspace, tenancyv1alpha1.WorkspaceContentClean) {
+						return false, fmt.Errorf("expect true status for condition %s", tenancyv1alpha1.WorkspaceContentClean)
+					}
+
+					return true, nil
+				}, 30*time.Second)
```
Comment (sttts) [inline test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:128]: `wait.ForeverTestTimeout`

File: test/e2e/reconciler/clusterworkspacedeletion/controller_test.go [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+
+				// finally check if all resources has been removed.
+				nslist, err := kubeClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+				require.NoError(t, err, "failed to list namespaces in workspace %s", workspace.Name)
+				require.Equal(t, 0, len(nslist.Items))
+
+				cmlist, err := kubeClient.CoreV1().ConfigMaps(metav1.NamespaceAll).List(ctx, metav1.ListOptions{})
+				require.NoError(t, err, "failed to list configmaps in workspace %s", workspace.Name)
+				require.Equal(t, 0, len(cmlist.Items))
+			},
```
Comment (sttts) [inline test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:155]: would like to see another cluster-scoped resource in the test.

Comment (sttts) [reply test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:155]: Also add a second test case where an org workspace is deleted. That has the obvious complexity of nested deletions, but it also has virtual resources (`Workspace` which is mapped into the org workspace, but is actually served by the virtual workspace apiserver).

Comment (sttts) [reply test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:155]: cc @davidfestal – interesting topic when we map in virtual resources that are marked writeable in the fake CRD we put into the workspace, but readonly in the virtual workspace.

Comment (qiujian16) [reply test/e2e/reconciler/clusterworkspacedeletion/controller_test.go:172]: also create a ns in the workspace, ensuring it is deleted and another test to delete org workspace.

File: pkg/admission/namespacelifecycle/admission.go [hunk 22]
```diff
@@ -92,7 +93,9 @@ func (l *workspaceNamespaceLifecycle) Admit(ctx context.Context, a admission.Att
 	// call legacy namespace lifecycle at first
 	admissionErr := l.legacyNamespaceLifecycle.Admit(ctx, a, o)
 
-	// Only handle the case when namespace deletion is forbidden
+	// when the workspace is being deleted, we want to allow deletion of all workspaces. Hence, after
```
Comment (sttts) [inline pkg/admission/namespacelifecycle/admission.go:96]: "all namespaces". Maybe my fault in the suggestion :)

File: pkg/admission/plugins.go [hunk 23]
```diff
@@ -80,17 +80,17 @@ func RegisterAllKcpAdmissionPlugins(plugins *admission.Plugins) {
...
 
 var defaultOnPluginsInKcp = sets.NewString(
-	kcpnamespacelifecycle.PluginName,  // NamespaceLifecycle
-	limitranger.PluginName,            // LimitRanger
-	mutatingwebhook.PluginName,        // MutatingAdmissionWebhook
-	validatingwebhook.PluginName,      // ValidatingAdmissionWebhook
-	certapproval.PluginName,           // CertificateApproval
-	certsigning.PluginName,            // CertificateSigning
-	certsubjectrestriction.PluginName, // CertificateSubjectRestriction
+	workspacenamespacelifecycle.PluginName, // NamespaceLifecycle
```
Comment (sttts) [inline pkg/admission/plugins.go:87]: it's still called `NamespaceLifecycle`?

File: pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go [hunk 24]
```diff
@@ -205,7 +209,7 @@ func (c *Controller) process(ctx context.Context, key string) error {
 }
 
 func (c *Controller) patchStatus(ctx context.Context, old, new *tenancyv1alpha1.ClusterWorkspace) error {
-	if !equality.Semantic.DeepEqual(old.Status.Conditions, new.Status.Conditions) {
+	if equality.Semantic.DeepEqual(old.Status.Conditions, new.Status.Conditions) {
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go:212]: with this it actually became `patchConditions`, no?

**Review by sttts — APPROVED** [review]

**Pull request merged**